From c84ad446d9edbc3b14f9c3b3b64374ecb49fc4de Mon Sep 17 00:00:00 2001
From: Kenneth Graunke <kenneth@whitecape.org>
Date: Thu, 20 May 2021 00:14:53 -0700
Subject: [PATCH 06/20] iris: Use staging blits for reads from uncached
 buffers.

If we're doing CPU reads of a resource that doesn't have CPU caches
enabled for the mapping (say, in device local memory, or WC mapped),
then blit it to a temporary that does have those caches enabled.

Acked-by: Lionel Landwerlin <lionel.g.landwerlin@intel.com>
Part-of: <https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/10941>

diff --git a/src/gallium/drivers/iris/iris_resource.c b/src/gallium/drivers/iris/iris_resource.c
index 9adef05255f..0f91772a1e0 100644
--- a/src/gallium/drivers/iris/iris_resource.c
+++ b/src/gallium/drivers/iris/iris_resource.c
@@ -2000,8 +2000,12 @@ iris_transfer_map(struct pipe_context *ctx,
    if (fmtl->txc == ISL_TXC_ASTC)
       usage |= PIPE_MAP_DIRECTLY;
 
+   /* We can map directly if it wouldn't stall, there's no compression,
+    * and we aren't doing an uncached read.
+    */
    if (!map_would_stall &&
-       !isl_aux_usage_has_compression(res->aux.usage)) {
+       !isl_aux_usage_has_compression(res->aux.usage) &&
+       !((usage & PIPE_MAP_READ) && !res->bo->cache_coherent)) {
       usage |= PIPE_MAP_DIRECTLY;
    }
 
-- 
2.36.1

